sdk/identity/azure-identity/src/client_certificate_credential.cpp

```diff
@@ -100,50 +384,19 @@ ClientCertificateCredential::ClientCertificateCredential(
   std::string thumbprintBase64Str;
 
   {
-    std::vector<unsigned char> mdVec(EVP_MAX_MD_SIZE);
+    CertificateThumbprint mdVec;
+    try
     {
-      UniqueHandle<X509> x509;
-      {
-        // Open certificate file, then get private key and X509:
-        UniqueHandle<BIO> bio(BIO_new_file(clientCertificatePath.c_str(), "r"));
-        if (!bio)
-        {
-          throw AuthenticationException("Failed to open certificate file.");
-        }
-
-        m_pkey.reset(PEM_read_bio_PrivateKey(bio.get(), nullptr, nullptr, nullptr));
-        if (!m_pkey)
-        {
-          throw AuthenticationException("Failed to read certificate private key.");
-        }
-
-        x509.reset(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr));
-        if (!x509)
-        {
-          static_cast<void>(BIO_seek(bio.get(), 0));
-          x509.reset(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr));
-          if (!x509)
-          {
-            throw AuthenticationException("Failed to read X509 section.");
-          }
-        }
-      }
-
-      // Get certificate thumbprint:
-      unsigned int mdLen = 0;
-      const auto digestResult = X509_digest(x509.get(), EVP_sha1(), mdVec.data(), &mdLen);
-
-      if (!digestResult)
-      {
-        throw AuthenticationException("Failed to get certificate thumbprint.");
-      }
-
-      // Drop unused buffer space:
-      const auto mdLenSz = static_cast<decltype(mdVec)::size_type>(mdLen);
-      if (mdVec.size() > mdLenSz)
-      {
-        mdVec.resize(mdLenSz);
-      }
+      std::tie(mdVec, m_pkey) = ReadPemCertificate(clientCertificatePath);
+    }
+    catch (AuthenticationException&)
+    {
+      throw;
+    }
+    catch (std::exception& e)
+    {
+      // WIL does not throw AuthenticationException.
+      throw AuthenticationException(e.what());
```

Comment:
Do you want to consider catching the wil status exception here and handling it explicitly? Or just rely on the `e.what()`?

Reply:
I don't see a problem with `e.what()`. It provides a nice error message and the exception handler remains oblivious to WIL.

Reply:
That's fair. There is additional potentially actionable information in the wil exceptions, but your point is valid.